packages/server/src/studentTaskProgress/studentTaskProgress.controller.ts

```diff
@@ -0,0 +1,119 @@
+import { CourseRolesGuard } from 'guards/course-roles.guard';
+import { QueueRolesGuard } from 'guards/queue-role.guard';
+
+@Controller('studentTaskProgress')
+@UseGuards(JwtAuthGuard, CourseRolesGuard)
+@UseInterceptors(ClassSerializerInterceptor)
+export class StudentTaskProgressController {
+  constructor(private connection: Connection) {}
+  // TODO: make it so students can only retrieve their own taskProgress
+  // maybe make 3 endpoints. getMyAssignmentProgress (using @UserId), getStudentAssignmentProgress (only usable by TAs. Though, maybe instead just ship the assignment progress with the task question itself), and getAllAssignmentProgressForQueue (also only used by TAs. This can be called initially to save a lot of the initial calls.), can also have getAllAssignmentProgressForCourse
+  // will probably want to move the method to its own controller at that point (with a little note that mentions that the setting of taskProgress happens in the updateQuestion endpoint in the question controller).
+  @Get('student/:userId/:courseId/:assignmentName')
+  @Roles(Role.STUDENT, Role.TA, Role.PROFESSOR)
+  async getStudentAssignmentProgress(
+    @Param('userId') userId: number,
+    @Param('courseId') courseId: number,
+    @Param('assignmentName') assignmentName: string,
+  ): Promise<StudentAssignmentProgress | null> {
+    const studentTaskProgress = await StudentTaskProgressModel.findOne({
+      where: {
+        uid: userId,
+        cid: courseId,
+      },
+    });
+
+    if (
+      !studentTaskProgress ||
+      !studentTaskProgress.taskProgress ||
+      !studentTaskProgress.taskProgress[assignmentName]
+    ) {
+      return null;
+    }
+
+    const studentAssignmentProgress =
+      studentTaskProgress.taskProgress[assignmentName].assignmentProgress;
+
+    if (studentAssignmentProgress === undefined) {
+      return null;
+    }
+
+    return studentAssignmentProgress;
+  }
+
+  /**
+   * Retrieves the assignment progress for all students in a specific queue.
+   *
+   * It's probably a costly endpoint to call, thus don't call it often on the frontend.
+   *
+   * Getting only the studentAssignmentProgress for a specific queue is needed since some queues may have the same assignment loaded in them.
+   */
+  @Get('queue/:queueId/:courseId/:assignmentName')
+  @UseGuards(QueueRolesGuard)
+  @Roles(Role.TA, Role.PROFESSOR)
+  async getAllAssignmentProgressForQueue(
+    @Param('queueId') queueId: number,
+    @Param('courseId') courseId: number,
+    @Param('assignmentName') assignmentName: string,
+  ): Promise<AllStudentAssignmentProgress> {
+    // this returns the entire StudentTaskProgress for all students (which includes the assignment progress for stuff we don't care about)
+    // TODO: figure out how to only select the needed columns from the join with user. Every way of adding the select statement causes a syntax error.
+    const allStudentTaskProgressForQueue = await getRepository(
```

Comment:
Oh i finally get it now. So controllers are just supposed to be small files for handling the HTTP requests and the service files are supposed to be doing all the database interactions and logic. 

And I guess this whole repo uses controllers as whole endpoints and then just uses services for common functions/abstractions. I now see how this is so cursed lol

I guess I was just sorta copying the code i saw in other files and went with what i saw lol 

Maybe when I add more endpoints to this i'll do a refactor and move some stuff into a service file. Or if there's any duplicate logic i'll be sure to refactor it out into a service. For now, since this logic is the only place it is being used, I think it's fine to leave it as is